### PR TITLE
[HOPSWORKS-733] Refactor JobService to use JAXB and SparkConfiguration

### PR DESCRIPTION
**Pull request by robzor92**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [ ] Adds tests for the submitted changes (for bug fixes & features)
- [x] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by robzor92** [comment]
@tkakantousis @SirOibaf please review.

**Comment by robzor92** [comment]
https://github.com/logicalclocks/hops-testing/pull/201

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/JobController.java [hunk 1]
```diff
@@ -68,8 +70,19 @@ public Jobs createJob(Users user, Project project, JobConfiguration config) thro
...
     }
-    
-    Jobs created = jobFacade.create(user, project, config);
+
+    Jobs created;
+    try {
+      created = jobFacade.create(user, project, config);
+    } catch(RuntimeException re) {
+      if(re.getCause() instanceof JAXBException) {
+        throw new JobException(RESTCodes.JobErrorCode.JOB_CONFIGURATION_CONVERT_TO_JSON_ERROR, Level.SEVERE,
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/JobController.java]: @robzor92 when we catch exception and then throw RESTExceptions, must also set the dev message and the throwable in the RESTException constructor. In this case it should be 
`throw new JobException(RESTCodes.JobErrorCode.JOB_CONFIGURATION_CONVERT_TO_JSON_ERROR, Level.SEVERE, "Unable to create json from JobConfiguration", re.getMessage(), re);`

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+  @Override
+  public JobConfiguration convertToEntityAttribute(String jsonConfig) {
+    if (Strings.isNullOrEmpty(jsonConfig)) {
+      return null;
+    }
+    try {
+      JSONObject obj = new JSONObject(jsonConfig);
+      String jType = (String)obj.get("jobType");
+      JobType type = JobType.valueOf(jType);
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: @robzor92 you can do this in one line as String `jType` and then JobType `type` is a bit confusing. For example `JobType.valueOf(JobType.valueOf(jType);`

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      JAXBContext jc = createJAXBContext(type);
+      return unmarshal(jsonConfig, type, jc);
+    } catch (JAXBException e) {
+      throw new RuntimeException(e);
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: @robzor92 what's the motivation for catching the JAXBException and throwing RuntimeException?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+  private JAXBContext createJAXBContext(JobType jobType) throws JAXBException {
+    switch(jobType) {
+      case SPARK:
+      case PYSPARK:
+        return JAXBContextFactory.createContext(new Class[] {SparkJobConfiguration.class}, null);
+      case FLINK:
+        return JAXBContextFactory.createContext(new Class[] {FlinkJobConfiguration.class}, null);
+      default:
+        throw new JAXBException("Could not find a mapping for JobType " + jobType);
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: @robzor92 I suggest replacing JAXBException with IllegalArgumentException.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    unmarshaller.setProperty(MarshallerProperties.JSON_INCLUDE_ROOT, false);
+    unmarshaller.setProperty(MarshallerProperties.MEDIA_TYPE, MediaType.APPLICATION_JSON);
+    switch(jobType) {
+      case SPARK:
+      case PYSPARK:
+        return unmarshaller.unmarshal(json, SparkJobConfiguration.class).getValue();
+      case FLINK:
+        return unmarshaller.unmarshal(json, FlinkJobConfiguration.class).getValue();
+      default:
+        throw new JAXBException("Could not find a mapping for JobType " + jobType);
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: @robzor92 I suggest replacing JAXBException with IllegalArgumentException.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java (hunk 4 above)
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: Sounds good

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java (hunk 5 above)
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: Sounds good

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java (hunk 3 above)
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/configuration/JobConfigurationConverter.java]: Switch to using IllegalStateException, can't override method signature which is why this is being thrown.

**Comment by robzor92** [comment]
@tkakantousis @SirOibaf expat migration code finished https://github.com/logicalclocks/expat/pull/5/files

**Pull request merged**